### PR TITLE
Redo timestamp file change to allow partial rebuilds.

### PR DESCRIPTION
**Pull request by bbowman**
Fix previous implementation of creating the timestamp file at the beginning by
being more careful with project references between nuproj files and projects
pretending to be .nuprojs. The new implementation ensures that a .nuproj<->.nuproj
project reference is only evaluated in one configuration so that the timestamp file
creation doesn't happen more than once (happens once per conifg of .nuproj being built).

The full explanation of what was occuring is as follows:
```
A.nuproj -> B.nuproj -> NuprojInit.msbuildproj
         |--------------^

In the Any CPU config the build order was as follows:

A.nuproj (Any CPU) -> B.nuproj (Any CPU) -> NuprojInit.msbuildproj (Any CPU)
   |                     |
   |                     |-> B.nuproj (x86)
   |                     |-> B.nuproj (ARM)
   |
   |--> A.nuproj (x86)--> B.nuproj (x86) -> NuprojInit.msbuildproj (x86)
   |--> A.nuproj (ARM)--> B.nuproj (ARM) -> NuprojInit.msbuildproj (ARM)
```
The new build order is:
```
A.nuproj (Any CPU) -> B.nuproj (Any CPU) -> NuprojInit.msbuildproj (Any CPU)
   |                     |
   |                     |-> B.nuproj (x86)
   |                     |-> B.nuproj (ARM)
   |
   |--> A.nuproj (x86)
   |--> A.nuproj (ARM)
```
The key difference being NuprojInit.msbuildproj only being built in the original
configuration. An additional benefit of this change is that B.nuproj will be rebuilt less.

### CODE REVIEW

File: build/WinObjC.Frameworks/WinObjC.Frameworks.nuproj [hunk 1]
```diff
@@ -110,6 +110,9 @@
     <ProjectReference Include="$(MSBuildThisFileDirectory)..\UIKit.Xaml\dll\UIKit.Xaml.vcxproj" />
     <ProjectReference Include="$(MSBuildThisFileDirectory)..\UIKit\dll\UIKit.vcxproj" />
     <ProjectReference Include="$(MSBuildThisFileDirectory)..\WebKit\dll\WebKit.vcxproj" />
+
+  <!-- NuProj Project References. Ensure that these only get evaluated once by checking that this isn't the second time through this project -->
+  <ItemGroup Condition="'$(SkipGetPackageContentsForOtherPlatforms)' != 'true'">
```
Comment (memontic-ms) [inline build/WinObjC.Frameworks/WinObjC.Frameworks.nuproj]: ><ItemGroup Condition="'$(SkipGetPackageContentsForOtherPlatforms)' != 'true'"> [](start = 2, length = 78)

So this check is needed anytime a nuproj has a ProjectReference on another nuproj? As in, we will need to make sure to edit the .nuproj file and add this check whenever we add a reference to another nuproj?
Additionally, when does SkipGetPackageContentsForOtherPlatforms get set?

Comment (bbowman) [inline build/WinObjC.Frameworks/WinObjC.Frameworks.nuproj]: Correct. It gets set [here](https://github.com/Microsoft/WinObjC/blob/develop/common/winobjc.nuproj.common.targets#L17) which is when the .nuproj gets re-evaluated for the other configs.

File: tools/legacy/BuildMonitor/Legacy_BuildMonitor.csproj [hunk 2]
```diff
@@ -38,6 +38,7 @@
     <IncludeDebugSymbolsInLocalVSIXDeployment>true</IncludeDebugSymbolsInLocalVSIXDeployment>
     <CopyBuildOutputToOutputDirectory>true</CopyBuildOutputToOutputDirectory>
     <CopyOutputSymbolsToOutputDirectory>false</CopyOutputSymbolsToOutputDirectory>
+    <DeployExtension>False</DeployExtension>
```
Comment (rajsesh-msft) [inline tools/legacy/BuildMonitor/Legacy_BuildMonitor.csproj]: Did these just sneak in from a different change?

Comment (rajsesh-msft) [inline tools/legacy/BuildMonitor/Legacy_BuildMonitor.csproj]: never mind, found what happened here.


---
In reply to: [107792278](https://github.com/Microsoft/WinObjC/pull/2311#discussion_r107792278) [](ancestors = 107792278)

Comment (bbowman) [inline tools/legacy/BuildMonitor/Legacy_BuildMonitor.csproj]: these were in the in original. They are coming back from reverting the revert. I was seeing some instability in tools.sln builds without this. @memontic-ms says I don't these now but they don't hurt anything to my knowledge so I left them.

**Pull request merged**